First batch for the new components doc (#24571)
DataDog/datadog-agent: docs/public/components/faq.md

# FAQ

> TODO: Should we migrate the FAQ from confluence here ?

Reviewer: note todo

## Optional dependency

You might need to express the fact that some of your dependencies are optional. This often happens for

> components that interact with many other components **if available** (ie: were included at compile time). This allow

Reviewer: ```suggestion
components that interact with many other components **if available** (that is, if they were included at compile time). This allows
```

> components that interact with many other components **if available** (that is, if they were included at compile time). This allows

Reviewer 2: This happens when a component can be enabled or not at runtime during the creation of the component. This component also have to expose methods.

> your component to interact with other without forcing their inclusion in the current binary.

Reviewer 3: ```suggestion
your component to interact with each other without forcing their inclusion in the current binary.
```

> The [optional.Option](https://github.com/DataDog/datadog-agent/tree/main/pkg/util/optional) type answer such need.

Reviewer: ```suggestion
The [optional.Option](https://github.com/DataDog/datadog-agent/tree/main/pkg/util/optional) type answers this need.
```

> A good example would our metadata components that are included in multiple binaries (`core-agent`, `DogStatsD`, ...).

Reviewer: ```suggestion
For examples, consider the metadata components that are included in multiple binaries (`core-agent`, `DogStatsD`, etc.).
```

> For examples, consider the metadata components that are included in multiple binaries (`core-agent`, `DogStatsD`, etc.).

Reviewer 2: ```suggestion
For example, consider the metadata components that are included in multiple binaries (`core-agent`, `DogStatsD`, etc.).
```

> Such components want to use the `sysprobeconfig` component if available. `sysprobeconfig` is available in the

Reviewer: ```suggestion
These components use the `sysprobeconfig` component if it is available. `sysprobeconfig` is available in the
```

`core-agent` but not in `DogStatsD`.

> To do this the `metadata` component will do this:

Reviewer: ```suggestion
To do this in the `metadata` component:
```

```
type Requires struct {
    SysprobeConf optional.Option[sysprobeconfig.Component]
    [...]
}

func NewMetadata(deps Requires) (metadata.Component) {

>     if sysprobeConf, available := deps.SysprobeConf.Get(); available {

Reviewer 3: Shorter, easier to type, used in some other places in go code:

```suggestion
    if sysprobeConf, found := deps.SysprobeConf.Get(); found {
```

        // interact with sysprobeconfig
    }
}
```

> The above code would produce a generique component, included in both `core-agent` and `DogStatsD` binaries, that **can**

Reviewer: ```suggestion
The above code would produce a generique component, included in both `core-agent` and `DogStatsD` binaries, that **can**
```
```suggestion
The above code produces a generic component, included in both `core-agent` and `DogStatsD` binaries, that **can**
```

interact with `sysprobeconfig` without forcing the binaries to compile with it.

> This pattern can be used for every components since they all provide a convertion function to `Fx` to convert their

Reviewer: ```suggestion
You can use this pattern for every component, since all components provide Fx with a conversion function to convert their
```
